polls: Add option for modal to create polls.

N-Shar-ma opened the pull request:
Earlier the `/poll` slash command was the only way to create polls. To increase user friendliness with a GUI, a button to launch a modal to create a poll, has been added to the compose box. This button is enabled only when the compose box is empty, to avoid complexities with losing / having to save as draft any message already being composed.

The modal has a form which on submission frames a message using the `/poll` syntax and the data input in the form, and sets the content of the compose box to that message, which the user can then send.

Fixes: #20304.

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/68962290/de99cc66-3ef3-4cce-9039-7bd6b54c9325)
![image](https://github.com/zulip/zulip/assets/68962290/66d7fc97-2d07-4f9b-a26b-272d3a3a46e9)

The modal:
![image](https://github.com/zulip/zulip/assets/68962290/ae91c34c-c0e0-452f-b3e8-ba08d569dd4e)
After clicking the "Frame message" button in the modal it closes and the compose box is populated:
![image](https://github.com/zulip/zulip/assets/68962290/36465a74-92b1-49ef-85bb-dc0fb3fb10e6)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Nice! I left a few comments on the [CZO thread](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Modal.20to.20create.20polls/near/1638791).

karlstolley commented on web/templates/poll_modal_option.hbs:
```diff
@@ -0,0 +1,8 @@
+<li class='option-row'>
```
I'm surprised the linter didn't flag this, but let's stick with double-quotes around attribute values (and singles around the translation strings) :)

N-Shar-ma replied:
Understood the conventions, and fixed this template.

However I also observed few other instances of such violations in other templates which aren't being flagged. Just thought I'd bring that to notice

karlstolley replied:
Cool, well, you can join me on Team Consistency to try and fix those things as they pop up :)

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
+    overflow: hidden;
+
+    @media (width < $sm_min) {
+        height: 80vh;
```
Why not set a viewport-sensitive height like this everywhere, and not just the very tiniest mobile screens? I think it's also probably not the best idea to be letting viewport width determine an element's height.

N-Shar-ma replied:
I realized that this particular rule is actually a remnant from a previous iteration of the modal which had more content that would wrap at small widths, increasing its effecting content height.

Now in its current state, the fixed height works just fine at a standard mobile screen, so have removed this rule.

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
     }
 }
+
+#add-poll-modal {
```
Does it make better sense to put this in `modal.css`? I think our goal should be to have fewer component-specific things in `zulip.css`.

N-Shar-ma replied:
Seems like hot reloading doesn't work for `modal.css` -- I need to reload to have the changes apply :thinking:

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
+    }
+
+    .modal__content {
+        flex-grow: 1;
```
I can't see how this has an effect on anything in the modal, but then I think the use of `flex-direction: column` on most modals is unnecessary.

N-Shar-ma replied:
Right, this is redundant. Removed this rule, thanks for pointing this out!

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
+        .simplebar-content {
+            box-sizing: border-box;
+            height: 100%;
+            display: flex;
```
This works just fine as regular flow content; there's no need to set up a columnar flexbox that I can see.

N-Shar-ma replied:
Removing flex also solved the problem of the footer (button area) appearing too tall which @alya had pointed out in the [CZO thread](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Modal.20to.20create.20polls)! Thanks for the suggestion!

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
+
+    #add-poll-form {
+        display: flex;
+        flex-direction: column;
```
Strictly speaking, this columnar flex seems necessary. But you can still get by without the one I mentioned above, I think.

N-Shar-ma replied:
Yes, this one needs to be retained since without it, the scrollable list of options won't work

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
+        & input[type="text"] {
+            border: 1px solid hsl(0deg 0% 0% / 20%);
+            outline: 0;
+            height: 1.8rem;
```
It's good practice to document where values like these are coming from.

N-Shar-ma replied:
The border is from the compose textarea and the topic input (from `compose.css` which is dedicated to the compose area so I couldn't just add this selector for this rule in that file) and the height I honestly just eyeballed since the default seemed a little too cramped.

Should I mention this in a couple comments here?

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
+            overflow: auto;
+            flex-grow: 1;
+
+            .option-row {
```
Could the option rows be made into flexboxes? That might reduce some of the complexity here with various bits of margin, or the relative positioning on the elipsis.

N-Shar-ma replied:
Thanks for the idea! This did reduce the complexity, and options (and the question input) taking up all available space looks fine too

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
+                }
+
+                & input {
+                    width: clamp(190px, 50%, 300px);
```
While I love clamp, I don't quite understand its purpose here, since there's only a tiny slice of the very smallest mobile screens where its active. And again, I think if each row were a flex container, you could let flexbox be in charge of assigning available space.

N-Shar-ma replied:
Did just that, let them take up all available space in a flexbox

karlstolley commented on web/styles/zulip.css:
```diff
@@ -3295,3 +3295,86 @@ select.invite-as {
+            }
+
+            .option-row:first-child {
+                margin-top: 0;
```
Why not just use `margin-bottom` to separate these? It's probably fine to have some extra margin at the bottom, and then you can save yourself this first-child selector.

N-Shar-ma replied:
I did consider using `margin-bottom`, however unless we remove it from the last row (which is analogous to this current code) it tends to cause an unnecessary scrollbar when there are 4 options, since all for rows fit visually, but the last option's bottom margin doesn't and causes the scrollbar to appear

karlstolley commented on web/templates/poll_modal_option.hbs:
```diff
@@ -0,0 +1,8 @@
+<li class='option-row'>
+    <i class="fa fa-ellipsis-v" aria-hidden="true"></i>
+    <i class="fa fa-ellipsis-v" aria-hidden="true"></i>
```
Rather than these doubled-up icons, which don't look quite right to me, we should see if @terpimost has a similar drag-handler icon that would work here.

N-Shar-ma replied:
Have changed to using the [suggested single icon](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Modal.20to.20create.20polls/near/1641737), and here's how it looks now:
![image](https://github.com/zulip/zulip/assets/68962290/715d8c98-7be3-41f2-9e79-3f5eb6cf72ed)

N-Shar-ma commented:
@karlstolley Apart from addressing the comments above, I have added another experimental commit which:
- Makes it so that trying to submit the form in the modal without a question being input leads to an error message like this: 
![image](https://github.com/zulip/zulip/assets/68962290/1b218604-0a1a-4916-9812-76edda72f5d0)
- Makes it impossible to drag the last empty row, also hiding it's drag handle icon, and making the cursor normal:
![image](https://github.com/zulip/zulip/assets/68962290/1f3998dc-19c9-4340-8091-e2eb9f4b1d0f)

karlstolley commented:
This is looking and working pretty good, @N-Shar-ma. It's a little weird that the gray border above the buttons shows up along with the error message (which itself seems useful to me).

Otherwise, I don't have much else to add here.

N-Shar-ma commented:
@karlstolley Thanks for the review! The gray border shows up on addition of the error message as then the modal's body content becomes too much too fit in vertically and hence it becomes scrollable, which is the condition for showing the gray border. But yes, it showing up abruptly does seem a little odd.

What do you suggest are the next steps for this PR?

karlstolley commented:
Weird. I would expect that same border then when adding more poll responses than could be displayed, but I didn't see it under that condition. Is that expected?

N-Shar-ma commented:
@karlstolley Though weird here, this behaviour is expected. Alya too had raised the same concern on CZO, and [here's](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Modal.20to.20create.20polls/near/1639490) my reply elaborating on this.

timabbott commented on web/src/compose_setup.js:
```diff
@@ -70,6 +76,11 @@ export function initialize() {
         } else {
             $("#compose_close").attr("data-tooltip-template-id", "compose_close_tooltip_template");
         }
+        if (compose_text_length > 0) {
```
Let's add a blank line and then comment line just before this, noting that "The poll widget requires an empty compose box."

timabbott commented on web/src/poll_modal.js:
```diff
@@ -0,0 +1,56 @@
+    }
+    const $options_div = $(e.target).parent().parent();
+    create_option_row($options_div);
+}
```
Would it make sense to have `$(e.target)` be the parameter for this function, maybe with a better variable name, like `$foo_element`?

timabbott commented on web/src/poll_modal.js:
```diff
@@ -0,0 +1,56 @@
+}
+
+function delete_option_row(e) {
+    const $row = $(e.currentTarget).parent();
```
Should we be using `.closest(".classname")` rather than these `.parent()` references? Usually that's a better plan, since it's less sensitive to minor refactoring of the HTML.

timabbott commented on web/src/poll_modal.js:
```diff
@@ -0,0 +1,56 @@
+export function poll_options_setup() {
+    const $poll_options_list = $("#add-poll-form .poll-options-list");
+
+    $poll_options_list.on("input", ".option-row input", add_option_row);
```
Does the input have a class such that it could be `input.class-name` here? Generally that's a better selector style.

timabbott commented on web/styles/modal.css:
```diff
@@ -383,3 +383,81 @@
+#add-poll-modal {
+    /* this height allows 3-4 option rows
+    to fit in without need for scrolling */
+    height: 450px;
```
Does this behave reasonably in not-very-tall windows?

N-Shar-ma replied:
Yes; in window heights shorter than 450px too, the entire modal fits inside and the scrollable options section shrinks but remains scrollable

timabbott commented on web/styles/modal.css:
```diff
@@ -383,3 +383,81 @@
+        overflow: hidden;
+        height: 100%;
+
+        & label {
```
Better to attach a class to both this label and the `input` below. 

See, for example, https://chat.zulip.org/#narrow/stream/6-frontend/topic/CSS.20optimization/near/1665704.

timabbott commented:
@alya are you happy with the UX experience of this? If so, we could move it onto #26783 once the above selector comments are resolved, since it'll have the same integration challenges as the other compose buttons.

N-Shar-ma commented:
@timabbott Thanks for the feedback; have addressed all of it!

alya commented:
> @alya are you happy with the UX experience of this? If so, we could move it onto https://github.com/zulip/zulip/pull/26783 once the above selector comments are resolved, since it'll have the same integration challenges as the other compose buttons.

My only nit is that we should add a period after the "Please enter a question" error message.

Also, it would actually better to disable the "Add poll" button while the question field is empty, with a "Please enter a question." tooltip in that state. (We have other modals where we do this sort of thing.)

N-Shar-ma commented:
> Also, it would actually better to disable the "Add poll" button while the question field is empty, with a "Please enter a question." tooltip in that state. (We have other modals where we do this sort of thing.)

@alya Could you please give an example of a modal with such a button?

N-Shar-ma commented:
@alya Done! Addressed the nit too, though now the error message doesn't really ever show up since the form submission button is disabled when the question is empty, and it just acts as a backup safeguard.
![image](https://github.com/zulip/zulip/assets/68962290/add27037-56f6-4b94-93dc-d6e45cc0266f)

alya commented:
Looks good to me in manual testing! So I think it should be reading for getting moved onto https://github.com/zulip/zulip/pull/26783.

N-Shar-ma commented:
@timabbott Should I simply rebase these 2 commits onto #26783, and close this PR?

timabbott commented:
Yes, that sounds right to me!
